fendermint/vm/interpreter/src/fvm/state/genesis.rs

```diff
@@ -99,25 +100,35 @@ where
     pub async fn new(
         store: DB,
         multi_engine: Arc<MultiEngine>,
-        bundle: &[u8],
+        builtin_actor_bundle: &[u8],
         custom_actor_bundle: &[u8],
     ) -> anyhow::Result<Self> {
-        // Load the builtin actor bundle.
-        let (manifest_version, manifest_data_cid): (u32, Cid) =
-            parse_bundle(&store, bundle).await?;
-        let manifest = Manifest::load(&store, &manifest_data_cid, manifest_version)?;
-
         // Load the custom actor bundle.
         let (custom_manifest_version, custom_manifest_data_cid): (u32, Cid) =
             parse_bundle(&store, custom_actor_bundle).await?;
         let custom_actor_manifest =
             CustomActorManifest::load(&store, &custom_manifest_data_cid, custom_manifest_version)?;
 
+        let mut intermediate_builtin = load_builtin(builtin_actor_bundle).await?;
+
+        replace_built_in_actor(
+            fendermint_actor_eam::IPC_EAM_ACTOR_NAME,
+            &custom_actor_manifest,
+            &mut intermediate_builtin,
+        )?;
+
+        // copy builtin actors from memory to store
+        intermediate_builtin.mem_store.copy_to(&store)?;
```

Comment:
If you can do this here, I'm confident that you can also do this in the interpreter so that it can be offered to the users as well, and as I mentioned the actor name/CID can be kept together with the definition of its state, rather than be split in two places: that the name is replaced here, _and_ you have to follow up with the corresponding state in a completely different place, and if you don't then things go horribly wrong.

Reply:
I thought about doing what you mentioned, but the issue I found quite strange was the FvmGenesisState takes the manifest_cid and manifest as part of the struct. When we deploy the contracts, there is no access to the original builtin actor bundle . I understand where you are coming from and I think that’s the better way. But I feel we can refactor the FvmGenesisState to accept bundle_bytes in place of manifest_cid and manifest that way we can have everything in one place and it would feel natural.

Reply:
I might be wrong as I haven't tried it myself, but I don't think taking manifest bytes or a loaded and parsed manifest makes much of a difference to what I'm saying. The manifests are loaded in `new` and loaded immediately, so that methods like `create_builtin_actor` and `create_custom_actor` can refer to the content by code IDs and names. 

I think you are pointing out this difference because we are still talking about a different thing. I suspect you want to keep it as bytes so you can manipulate those bytes before loading them in the store, as you do now IIRC. I am suggesting not to do that. 

Instead, I suggest that we load all bundles into manifests at the start, as we do now. The builtin actor bundle goes in the ref-fvm Manifest, the custom actors go in our custom manifest that indexes by name. Then we load the regular builtin actors as we do now, no change in that. After that, we implement the code that can `replace_builtin_actor` method: when it's called, the `System` actor will already exist, and its `builtin_actors` CID will match the one in the `FvmGenesisState::manifest_data_cid`. At that point we can ignore the `manifest_data_cid`; we load the `Vec<(String, Cid)>` pointed at the by the `System` actors state, find the name we want to replace, store the new vector, assign its CID to the actor state, store the actor state, modify the system actor's ID in the `StateTree` to point at the new state, and we can also store the values inside the `FvmGenesisState` as well if we load a new `Manifest`. 

This leaves some intermediate values in the blockstore but that's okay because they won't be persisted as nothing will point at them.

Reply:
@aakoshh I see where you are coming from, so now `manifest_data_cid` is just for initial loading and subsequent updates are directly read from the system actor state. This works for me as well!